client/rest/src/main/java/org/elasticsearch/client/RestClient.java

```diff
@@ -361,11 +361,19 @@ public void cancelled() {
     private void setHeaders(HttpRequest httpRequest, Header[] requestHeaders) {
         Objects.requireNonNull(requestHeaders, "request headers must not be null");
         for (Header defaultHeader : defaultHeaders) {
-            httpRequest.setHeader(defaultHeader);
+            httpRequest.addHeader(defaultHeader);
         }
-        for (Header requestHeader : requestHeaders) {
-            Objects.requireNonNull(requestHeader, "request header must not be null");
-            httpRequest.setHeader(requestHeader);
+        if (requestHeaders.length != 0) {
+            final Set<String> uniqueNames = new HashSet<>();
+            for (Header requestHeader : requestHeaders) {
+                Objects.requireNonNull(requestHeader, "request header must not be null");
+                // overwrite any default header
+                if (uniqueNames.add(requestHeader.getName())) {
+                    httpRequest.setHeader(requestHeader);
+                } else {
+                    httpRequest.addHeader(requestHeader);
```

Comment:
I don't follow why this if is needed. Shouldn't we always use addHeader here? why do we need the set? how does it help with default headers?